amalgamate.mjs

```diff
@@ -0,0 +1,264 @@
+        {
+            includeLines: () => [
+                // Need to define `_GNU_SOURCE` the first time we `#include` these.
+                "#define _GNU_SOURCE",
+                "#include <pthread.h>",
+                "#include <sched.h>",
+                "#undef _GNU_SOURCE",
+            ],
+        },
+        ...commands
+            .filter(cmd => !filesWithErrors.has(cmd.relativePath)),
+    ]
+        .map(cmd => cmd.includeLines().join("\n"))
+        .join("\n\n")
+        + "\n"
+        ;
+
+    const linkFlags = [];
+
+    const cc = process.env.CC ?? "cc";
+
+    const amalgamated = {
+        ii: {
+            // A post-preprocessing file, just to check exactly what's included or not.
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath(".ii"), "-E", amalgamatedPath(".c")],
+            file: amalgamatedPath(".c"),
+        },
+        o: {
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath(".o"), "-c", amalgamatedPath(".c")],
+            file: amalgamatedPath(".c"),
+        },
+        exe: {
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath("exe"), amalgamatedPath("c"), ...linkFlags],
+            file: amalgamatedPath(".c"),
+        },
+    };
+
+    return {
+        amalgamated,
+        includes,
+    };
+}
+
+
+
+function toShellCommand(cmd) {
+    const command = cmd.arguments && cmd.arguments.join(" ") || cmd.command;
+    return `(cd "${cmd.directory}" && ${command})`;
+}
+
+async function runCommand(cmd, ...extraArgs) {
+    let [exe, ...args] = [...cmd.arguments, ...extraArgs];
+    try {
+        return await execFile(exe, args, {
+            cwd: cmd.directory,
+            maxBuffer: 1_000_000_000,
+        });
+    } catch (e) {
+        return e;
+    }
+}
+
+async function redefinitions({ amalgamated }) {
+    const { stderr } = await runCommand(amalgamated.o, "-fdiagnostics-format=json");
+    const errors = JSON.parse(stderr);
+    return errors
+        .filter(error => error.message.match(/^rede(fini|clara)tion of /))
+        .map(error => {
+            const start = error.message.lastIndexOf("‘");
+            const end = error.message.lastIndexOf("’");
+            const identifier = error.message
+                .slice(start + 1, end)
+                .replace(/^(struct|union|enum) /, "")
+                ;
+            const path = error.locations[0].caret.file;
+            return { identifier, path };
+        })
+        .reduce((groups, { identifier, path }) => {
+            (groups[path] = groups[path] ?? []).push(identifier);
+            return groups;
+        }, {})
+        ;
+}
+
+async function writeAmalgamatedCommands(stage) {
+    const commands = [stage.amalgamated.o];
+
+    await fsp.writeFile(amalgamatedPath(".c"), stage.includes);
+    await fsp.writeFile(amalgamatedPath("compile_commands.json"), JSON.stringify(commands, null, 4));
+}
+
+async function main() {
```

Comment:
Are we getting any value out of async here? It seems like each step blocks on the previous one.

Reply:
You can't really block in JS, so stuff like file writing is always `async`.  Perf doesn't matter, though, so it doesn't need any concurrency.